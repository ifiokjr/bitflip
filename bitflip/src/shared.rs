pub struct AppError(anyhow::Error);

// Tell axum how to convert `AppError` into a response.
#[cfg(feature = "ssr")]
impl axum::response::IntoResponse for AppError {
	fn into_response(self) -> axum::response::Response {
		(
			http::StatusCode::INTERNAL_SERVER_ERROR,
			format!("App error: {}", self.0),
		)
			.into_response()
	}
}

// This enables using `?` on functions that return `Result<_, anyhow::Error>` to
// turn them into `Result<_, AppError>`. That way you don't need to do that
// manually.
impl<E> From<E> for AppError
where
	E: Into<anyhow::Error>,
{
	fn from(err: E) -> Self {
		Self(err.into())
	}
}
