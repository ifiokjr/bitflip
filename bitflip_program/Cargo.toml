[package]
name = "bitflip_program"
version = "0.0.0"
edition = "2021"
description = "Created with Anchor"

[lints]
workspace = true

[lib]
crate-type = ["cdylib", "lib"]
name = "bitflip_program"

[features]
default = ["cpi"]
cpi = []
no-entrypoint = []
no-idl = []
no-log-ix-name = []
idl-build = ["anchor-lang/idl-build", "anchor-spl/idl-build"]
serde = ["dep:serde", "dep:serde_with"]
client = ["serde", "no-entrypoint", "dep:typed-builder"]

[dependencies]
anchor-lang = { workspace = true, features = ["interface-instructions", "init-if-needed"] }
anchor-spl = { workspace = true }
bitflags = { workspace = true }
bytemuck = { workspace = true }
const_str_to_pubkey = { workspace = true }
derive_more = { workspace = true, features = ["add", "add_assign"] }
serde = { workspace = true, features = ["derive"], optional = true }
serde-big-array = { workspace = true }
serde_with = { workspace = true, optional = true }
typed-builder = { workspace = true, optional = true }

[dev-dependencies]
anyhow = { workspace = true }
assert2 = { workspace = true }
insta = { workspace = true, features = ["ron", "yaml", "redactions"] }
rstest = { workspace = true }
test_utils = { workspace = true }
test_utils_solana = { workspace = true }
tokio = { workspace = true, features = ["test-util", "rt", "macros", "time"] }
