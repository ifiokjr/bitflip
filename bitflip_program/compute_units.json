{
  "instructions": {
    "config_initialize": {
      "compute_units": 17781,
      "rounded_compute_units": 20000,
      "description": "Initialize the config"
    },
    "config_update_authority": {
      "compute_units": 2464,
      "rounded_compute_units": 10000,
      "description": "Update the authority of the config"
    },
    "flip_bit_test": {
      "compute_units": 65231,
      "rounded_compute_units": 70000,
      "description": "Flip a single bit from `0` to `1`"
    },
    "game_initialize": {
      "compute_units": 13227,
      "rounded_compute_units": 20000,
      "description": "Initialize the game"
    },
    "game_refresh_signer": {
      "compute_units": 2371,
      "rounded_compute_units": 10000,
      "description": "Refresh the signer"
    },
    "token_group_initialize": {
      "compute_units": 47820,
      "rounded_compute_units": 50000,
      "description": "Initialize the token group"
    },
    "token_initialize:bit": {
      "compute_units": 63507,
      "rounded_compute_units": 70000,
      "description": "Initialize group token: Bit"
    },
    "token_initialize:gibibit": {
      "compute_units": 60069,
      "rounded_compute_units": 70000,
      "description": "Initialize member token: GibiBit"
    },
    "token_initialize:kibibit": {
      "compute_units": 60068,
      "rounded_compute_units": 70000,
      "description": "Initialize member token: KibiBit"
    },
    "token_initialize:mebibit": {
      "compute_units": 60065,
      "rounded_compute_units": 70000,
      "description": "Initialize member token: MebiBit"
    }
  },
  "version": "0.0.0"
}