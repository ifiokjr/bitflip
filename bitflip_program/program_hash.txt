50bd03d71a9fee65661b2bf98544cd5892ab8ff4ac6d8e5e1cfc186a2a7f6fa7
