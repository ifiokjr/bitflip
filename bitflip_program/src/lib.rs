mod constants;
pub mod cpi;
mod errors;
mod events;
#[cfg(feature = "client")]
mod instructions;
mod loaders;
mod pda;
mod processor;
mod state;
mod utils;

use steel::*;

pub use crate::constants::*;
pub use crate::errors::*;
pub use crate::events::*;
#[cfg(feature = "client")]
pub use crate::instructions::*;
pub use crate::loaders::*;
pub use crate::pda::*;
pub use crate::processor::*;
pub use crate::state::*;
pub use crate::utils::*;

solana_package_metadata::declare_id_with_package_metadata!("solana.program-id");

#[cfg(not(feature = "no-entrypoint"))]
entrypoint!(process_instruction);

// #[cfg(not(feature = "no-entrypoint"))]
solana_security_txt::security_txt! {
	// Required fields
	name: "Bitflip",
	project_url: "https://bitflip.art",
	contacts: "link:https://github.com/ifiokjr/bitflip/security/advisories/new,email:security@kickjump.co,discord:https://bitflip.art/discord",
	policy: "https://github.com/ifiokjr/bitflip/blob/main/security.md",

	// Optional Fields
	preferred_languages: "en",
	source_code: "https://github.com/ifiokjr/bitflip/tree/main/bitflip_program/",
	source_revision: default_env::default_env!("GITHUB_SHA", ""),
	source_release: default_env::default_env!("GITHUB_REF_NAME", ""),
	auditors: concat!("Verifier pubkey: ", default_env::default_env!("GITHUB_SHA", "")),
	encryption: "",
	acknowledgements: "Thank you to our bug bounty degens!"
}

#[cfg(test)]
pub(crate) fn leak<T>(value: T) -> &'static mut T {
	Box::leak(Box::new(value))
}
