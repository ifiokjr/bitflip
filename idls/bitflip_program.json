{
	"address": "5AuNvfV9Xi9gskJpW2qQJndQkFcwbWNV6fjaf2VvuEcM",
	"metadata": {
		"name": "bitflip_program",
		"version": "0.0.0",
		"spec": "0.1.0",
		"description": "Created with Anchor"
	},
	"instructions": [
		{
			"name": "initialize_bits_data_section",
			"docs": [
				"This will initialize a single section of the bits data. It should be",
				"called 16 times with the correct index of the data chunch."
			],
			"discriminator": [
				171,
				216,
				141,
				120,
				246,
				225,
				197,
				148
			],
			"accounts": [
				{
					"name": "config",
					"docs": [
						"The program configuration."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									99,
									111,
									110,
									102,
									105,
									103
								]
							}
						]
					}
				},
				{
					"name": "bits_meta",
					"docs": [
						"The meta data for the full bits state."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									115
								]
							},
							{
								"kind": "account",
								"path": "bits_meta.index",
								"account": "BitsMetaState"
							}
						]
					}
				},
				{
					"name": "bits_data_section",
					"docs": [
						"This is a section of the bits data being initialized."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									115
								]
							},
							{
								"kind": "account",
								"path": "config.bits_index",
								"account": "ConfigState"
							},
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									115,
									95,
									115,
									101,
									99,
									116,
									105,
									111,
									110
								]
							},
							{
								"kind": "arg",
								"path": "section"
							}
						]
					}
				},
				{
					"name": "authority",
					"docs": [
						"The authority that is able to sign for updates to the config and",
						"initiate new games."
					],
					"writable": true,
					"signer": true,
					"relations": [
						"config"
					]
				},
				{
					"name": "system_program",
					"docs": [
						"This is needed for initializing the bit state."
					],
					"address": "11111111111111111111111111111111"
				}
			],
			"args": [
				{
					"name": "section",
					"type": "u8"
				}
			]
		},
		{
			"name": "initialize_bits_meta",
			"docs": [
				"This will initialize the meta state for the bits."
			],
			"discriminator": [
				202,
				134,
				244,
				179,
				235,
				49,
				7,
				137
			],
			"accounts": [
				{
					"name": "config",
					"docs": [
						"The program configuration."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									99,
									111,
									110,
									102,
									105,
									103
								]
							}
						]
					}
				},
				{
					"name": "bits_meta",
					"docs": [
						"Can't initialize the full state at one time, must incrementally add the",
						"state."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									115
								]
							},
							{
								"kind": "account",
								"path": "config.bits_index",
								"account": "ConfigState"
							}
						]
					}
				},
				{
					"name": "authority",
					"docs": [
						"The authority that is able to sign for updates to the config and",
						"initiate new games."
					],
					"writable": true,
					"signer": true,
					"relations": [
						"config"
					]
				},
				{
					"name": "system_program",
					"docs": [
						"This is needed for initializing the bit state."
					],
					"address": "11111111111111111111111111111111"
				}
			],
			"args": []
		},
		{
			"name": "initialize_config",
			"docs": [
				"Initialize the configuration for the program."
			],
			"discriminator": [
				208,
				127,
				21,
				1,
				194,
				190,
				196,
				70
			],
			"accounts": [
				{
					"name": "config",
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									99,
									111,
									110,
									102,
									105,
									103
								]
							}
						]
					}
				},
				{
					"name": "admin",
					"docs": [
						"The hard-coded account that is used to initialize the program config",
						"once."
					],
					"writable": true,
					"signer": true,
					"address": "4z5X2suocz9szaQnSshj2AW8tuLgUVmYUxiW9hhPaRHs"
				},
				{
					"name": "treasury",
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									116,
									114,
									101,
									97,
									115,
									117,
									114,
									121
								]
							}
						]
					}
				},
				{
					"name": "system_program",
					"docs": [
						"This is needed for initializing the bit state."
					],
					"address": "11111111111111111111111111111111"
				}
			],
			"args": [
				{
					"name": "props",
					"type": {
						"defined": {
							"name": "InitializeConfigProps"
						}
					}
				}
			]
		},
		{
			"name": "initialize_token",
			"docs": [
				"Initialize the token account. This must be called before the first game",
				"starts to generate the reward token."
			],
			"discriminator": [
				38,
				209,
				150,
				50,
				190,
				117,
				16,
				54
			],
			"accounts": [
				{
					"name": "config",
					"docs": [
						"The program configuration."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									99,
									111,
									110,
									102,
									105,
									103
								]
							}
						]
					}
				},
				{
					"name": "authority",
					"docs": [
						"The program authority which must be a signer to create this token."
					],
					"writable": true,
					"signer": true,
					"relations": [
						"config"
					]
				},
				{
					"name": "mint",
					"writable": true
				},
				{
					"name": "treasury",
					"writable": true
				},
				{
					"name": "treasury_token_account",
					"writable": true
				},
				{
					"name": "associated_token_program"
				},
				{
					"name": "token_program"
				},
				{
					"name": "system_program"
				},
				{
					"name": "bitflip_program",
					"docs": [
						"The program that is for signing."
					],
					"address": "5AuNvfV9Xi9gskJpW2qQJndQkFcwbWNV6fjaf2VvuEcM"
				}
			],
			"args": [
				{
					"name": "props",
					"type": {
						"defined": {
							"name": "InitializeTokenProps"
						}
					}
				}
			]
		},
		{
			"name": "initialize_token_inner",
			"docs": [
				"INNER: This method can only be called by [`initialize_token`]. It uses",
				"the `treasury` signer as the authority for the mint account."
			],
			"discriminator": [
				112,
				121,
				158,
				27,
				165,
				27,
				94,
				26
			],
			"accounts": [
				{
					"name": "authority",
					"docs": [
						"CHECKED: checked in [`InitializeToken`] outer call."
					],
					"writable": true,
					"signer": true
				},
				{
					"name": "mint",
					"docs": [
						"The token mint account."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									109,
									105,
									110,
									116
								]
							}
						]
					}
				},
				{
					"name": "treasury",
					"docs": [
						"The treasury account."
					],
					"writable": true,
					"signer": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									116,
									114,
									101,
									97,
									115,
									117,
									114,
									121
								]
							}
						]
					}
				},
				{
					"name": "treasury_token_account",
					"docs": [
						"The associated token account for the treasury which will hold the minted",
						"tokens."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "account",
								"path": "treasury"
							},
							{
								"kind": "account",
								"path": "token_program"
							},
							{
								"kind": "account",
								"path": "mint"
							}
						],
						"program": {
							"kind": "const",
							"value": [
								140,
								151,
								37,
								143,
								78,
								36,
								137,
								241,
								187,
								61,
								16,
								41,
								20,
								142,
								13,
								131,
								11,
								90,
								19,
								153,
								218,
								255,
								16,
								132,
								4,
								142,
								123,
								216,
								219,
								233,
								248,
								89
							]
						}
					}
				},
				{
					"name": "associated_token_program",
					"docs": [
						"The program for associated tokens"
					],
					"address": "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL"
				},
				{
					"name": "token_program",
					"docs": [
						"The token program."
					],
					"address": "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb"
				},
				{
					"name": "system_program",
					"docs": [
						"Needed in case a reallocation is required for the project memory."
					],
					"address": "11111111111111111111111111111111"
				}
			],
			"args": [
				{
					"name": "props",
					"type": {
						"defined": {
							"name": "InitializeTokenInnerProps"
						}
					}
				}
			],
			"returns": "u8"
		},
		{
			"name": "set_bits",
			"docs": [
				"Flip bits based on the provided props."
			],
			"discriminator": [
				67,
				62,
				24,
				241,
				239,
				200,
				236,
				192
			],
			"accounts": [
				{
					"name": "config",
					"docs": [
						"The program configuration."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									99,
									111,
									110,
									102,
									105,
									103
								]
							}
						]
					}
				},
				{
					"name": "mint",
					"docs": [
						"The token mint account."
					],
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									109,
									105,
									110,
									116
								]
							}
						]
					}
				},
				{
					"name": "treasury",
					"docs": [
						"The treasury account which will transfer the spl tokens to the player."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									116,
									114,
									101,
									97,
									115,
									117,
									114,
									121
								]
							}
						]
					}
				},
				{
					"name": "treasury_token_account",
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "account",
								"path": "treasury"
							},
							{
								"kind": "account",
								"path": "token_program"
							},
							{
								"kind": "account",
								"path": "mint"
							}
						],
						"program": {
							"kind": "const",
							"value": [
								140,
								151,
								37,
								143,
								78,
								36,
								137,
								241,
								187,
								61,
								16,
								41,
								20,
								142,
								13,
								131,
								11,
								90,
								19,
								153,
								218,
								255,
								16,
								132,
								4,
								142,
								123,
								216,
								219,
								233,
								248,
								89
							]
						}
					}
				},
				{
					"name": "bits_meta",
					"docs": [
						"The meta data for the full bits state."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									115
								]
							},
							{
								"kind": "account",
								"path": "bits_meta.index",
								"account": "BitsMetaState"
							}
						]
					}
				},
				{
					"name": "bits_data_section",
					"docs": [
						"The data for this section of the bit canvas."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									115
								]
							},
							{
								"kind": "account",
								"path": "bits_meta.index",
								"account": "BitsMetaState"
							},
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									115,
									95,
									115,
									101,
									99,
									116,
									105,
									111,
									110
								]
							},
							{
								"kind": "arg",
								"path": "props.section"
							}
						]
					}
				},
				{
					"name": "player",
					"docs": [
						"The player of the bit games"
					],
					"writable": true,
					"signer": true
				},
				{
					"name": "player_token_account",
					"docs": [
						"The associated token account for the main authority."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "account",
								"path": "player"
							},
							{
								"kind": "account",
								"path": "token_program"
							},
							{
								"kind": "account",
								"path": "mint"
							}
						],
						"program": {
							"kind": "const",
							"value": [
								140,
								151,
								37,
								143,
								78,
								36,
								137,
								241,
								187,
								61,
								16,
								41,
								20,
								142,
								13,
								131,
								11,
								90,
								19,
								153,
								218,
								255,
								16,
								132,
								4,
								142,
								123,
								216,
								219,
								233,
								248,
								89
							]
						}
					}
				},
				{
					"name": "associated_token_program",
					"docs": [
						"The program for associated tokens"
					],
					"address": "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL"
				},
				{
					"name": "token_program",
					"docs": [
						"The token program."
					],
					"address": "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb"
				},
				{
					"name": "system_program",
					"docs": [
						"Needed for cpi payment instructions instructions."
					],
					"address": "11111111111111111111111111111111"
				}
			],
			"args": [
				{
					"name": "props",
					"type": {
						"defined": {
							"name": "SetBitsProps"
						}
					}
				}
			]
		},
		{
			"name": "start_bits_session",
			"docs": [
				"Start the bits session with and set the flipped bits to the correct",
				"number. This can only be called once 16 sections have been initialized."
			],
			"discriminator": [
				86,
				86,
				164,
				215,
				248,
				251,
				126,
				195
			],
			"accounts": [
				{
					"name": "config",
					"docs": [
						"The program configuration."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									99,
									111,
									110,
									102,
									105,
									103
								]
							}
						]
					}
				},
				{
					"name": "bits_meta",
					"docs": [
						"The meta data account for the game."
					],
					"writable": true,
					"pda": {
						"seeds": [
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									102,
									108,
									105,
									112
								]
							},
							{
								"kind": "const",
								"value": [
									98,
									105,
									116,
									115
								]
							},
							{
								"kind": "account",
								"path": "config.bits_index",
								"account": "ConfigState"
							}
						]
					}
				},
				{
					"name": "authority",
					"docs": [
						"The authority that is able to sign for updates to the config and",
						"initiate new games."
					],
					"writable": true,
					"signer": true,
					"relations": [
						"config"
					]
				},
				{
					"name": "system_program",
					"docs": [
						"This is needed for initializing the bit state."
					],
					"address": "11111111111111111111111111111111"
				}
			],
			"args": [
				{
					"name": "flipped_bits",
					"type": "u32"
				}
			]
		}
	],
	"accounts": [
		{
			"name": "BitsDataSectionState",
			"discriminator": [
				140,
				124,
				237,
				202,
				139,
				138,
				99,
				116
			]
		},
		{
			"name": "BitsMetaState",
			"discriminator": [
				93,
				244,
				217,
				115,
				235,
				203,
				72,
				85
			]
		},
		{
			"name": "ConfigState",
			"discriminator": [
				193,
				77,
				160,
				128,
				208,
				254,
				180,
				135
			]
		}
	],
	"errors": [
		{
			"code": 6000,
			"name": "BitsIncreaseSpaceInvalid",
			"msg": "The space is already fully initialized"
		},
		{
			"code": 6001,
			"name": "BitsUnchanged",
			"msg": "No update recorded"
		},
		{
			"code": 6002,
			"name": "InvalidAccount",
			"msg": "The provided account was invalid"
		},
		{
			"code": 6003,
			"name": "InvalidBitsLength",
			"msg": "The bits array is an invalid length"
		},
		{
			"code": 6004,
			"name": "InvalidBitsDataSection",
			"msg": "Invalid bit data section requested"
		},
		{
			"code": 6005,
			"name": "InvalidBitsDataSectionIndex",
			"msg": "Invalid bits data section index requested"
		},
		{
			"code": 6006,
			"name": "InvalidBitsDataSectionLength",
			"msg": "Invalid bits data section array length"
		},
		{
			"code": 6007,
			"name": "InvalidBitsDataSectionOrder",
			"msg": "Data sections initialized out of order"
		},
		{
			"code": 6008,
			"name": "InvalidFlippedBits",
			"msg": "An invalid number of flipped bits was provided"
		},
		{
			"code": 6009,
			"name": "Invalid256BitsDataSectionIndex",
			"msg": "The data section index must be a multiple of 16"
		},
		{
			"code": 6010,
			"name": "InvalidBitOffset",
			"msg": "The bit offset is invalid and must be less than 16"
		},
		{
			"code": 6011,
			"name": "NotRunning",
			"msg": "The current `BitsMetaState` is not running"
		},
		{
			"code": 6012,
			"name": "UnauthorizedAdmin",
			"msg": "The admin used was incorrect"
		},
		{
			"code": 6013,
			"name": "AllSectionsInitialized",
			"msg": "All bit data sections have already been initialized"
		},
		{
			"code": 6014,
			"name": "TokenNotInitialized",
			"msg": "The token is not yet initialized"
		}
	],
	"types": [
		{
			"name": "BitsDataSectionState",
			"docs": [
				"The data for each section of the the data. The total data is split into 16",
				"sections and this is one of the sections."
			],
			"serialization": "bytemuckunsafe",
			"repr": {
				"kind": "c"
			},
			"type": {
				"kind": "struct",
				"fields": [
					{
						"name": "data",
						"docs": [
							"The state of the bits that are represented as flippable bits on the",
							"frontend."
						],
						"type": {
							"array": [
								"u16",
								4096
							]
						}
					}
				]
			}
		},
		{
			"name": "BitsMetaState",
			"docs": [
				"Adding [`BitState::on`] to [`BitState::off`] should always equal `1_000_000`"
			],
			"type": {
				"kind": "struct",
				"fields": [
					{
						"name": "start_time",
						"docs": [
							"The start time. If 0 then it hasn't started yet. Using an `Option` here",
							"would waste an extra byte."
						],
						"type": "i64"
					},
					{
						"name": "flips",
						"docs": [
							"The number of bit flips that have occurred."
						],
						"type": "u64"
					},
					{
						"name": "on",
						"docs": [
							"The number of bits that are on."
						],
						"type": "u32"
					},
					{
						"name": "off",
						"docs": [
							"The number of bits that are off."
						],
						"type": "u32"
					},
					{
						"name": "index",
						"docs": [
							"The index of this currently played game."
						],
						"type": "u8"
					},
					{
						"name": "bump",
						"docs": [
							"The bump for this account."
						],
						"type": "u8"
					},
					{
						"name": "section_bumps",
						"docs": [
							"The bumps for each initialized section"
						],
						"type": "bytes"
					}
				]
			}
		},
		{
			"name": "ConfigState",
			"type": {
				"kind": "struct",
				"fields": [
					{
						"name": "authority",
						"docs": [
							"The authority which can update this config."
						],
						"type": "pubkey"
					},
					{
						"name": "lamports_per_bit",
						"docs": [
							"The number of lamports per bit change."
						],
						"type": "u64"
					},
					{
						"name": "bump",
						"docs": [
							"Store the bump to save compute units."
						],
						"type": "u8"
					},
					{
						"name": "treasury_bump",
						"docs": [
							"The treasury account bump where fees are sent and where the minted",
							"tokens are transferred."
						],
						"type": "u8"
					},
					{
						"name": "mint_bump",
						"docs": [
							"The mint account bump."
						],
						"type": "u8"
					},
					{
						"name": "bits_index",
						"docs": [
							"There will be a maximum of 4 games."
						],
						"type": "u8"
					}
				]
			}
		},
		{
			"name": "InitializeConfigProps",
			"type": {
				"kind": "struct",
				"fields": [
					{
						"name": "authority",
						"docs": [
							"The authority is the solana account which will have rights to update",
							"this configuration."
						],
						"type": "pubkey"
					}
				]
			}
		},
		{
			"name": "InitializeTokenInnerProps",
			"type": {
				"kind": "struct",
				"fields": [
					{
						"name": "name",
						"type": "string"
					},
					{
						"name": "symbol",
						"type": "string"
					},
					{
						"name": "uri",
						"type": "string"
					},
					{
						"name": "treasury_bump",
						"type": "u8"
					}
				]
			}
		},
		{
			"name": "InitializeTokenProps",
			"type": {
				"kind": "struct",
				"fields": [
					{
						"name": "name",
						"type": "string"
					},
					{
						"name": "symbol",
						"type": "string"
					},
					{
						"name": "uri",
						"type": "string"
					}
				]
			}
		},
		{
			"name": "SetBitsProps",
			"type": {
				"kind": "struct",
				"fields": [
					{
						"name": "section",
						"docs": [
							"The data section being updated."
						],
						"type": "u8"
					},
					{
						"name": "index",
						"docs": [
							"The index of the bit being set."
						],
						"type": "u16"
					},
					{
						"name": "variant",
						"docs": [
							"The new bit values."
						],
						"type": {
							"defined": {
								"name": "SetBitsVariant"
							}
						}
					}
				]
			}
		},
		{
			"name": "SetBitsVariant",
			"type": {
				"kind": "enum",
				"variants": [
					{
						"name": "On",
						"fields": [
							"u16"
						]
					},
					{
						"name": "Off",
						"fields": [
							"u16"
						]
					},
					{
						"name": "Bit16",
						"fields": [
							"u16"
						]
					},
					{
						"name": "Bits256",
						"fields": [
							{
								"vec": "u16"
							}
						]
					}
				]
			}
		}
	]
}
