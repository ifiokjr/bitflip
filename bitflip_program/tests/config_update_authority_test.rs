#![cfg(feature = "client")]

use std::future::Future;

use assert2::check;
use bitflip_program::config_update_authority;
use bitflip_program::get_pda_config;
use bitflip_program::BitflipError;
use bitflip_program::ConfigState;
use shared::create_authority_keypair;
use shared::create_config_accounts;
use shared::create_program_context_with_factory;
use shared::create_token_accounts;
use shared::ToRpcClient;
use solana_sdk::instruction::InstructionError;
use solana_sdk::signature::Keypair;
use solana_sdk::transaction::TransactionError;
use solana_sdk::transaction::VersionedTransaction;
use steel::*;
use test_utils::create_insta_redaction;
use test_utils_solana::prelude::*;

mod shared;

#[test_log::test(tokio::test)]
async fn config_update_authority_test() -> anyhow::Result<()> {
	shared_config_update_authority_test(create_banks_client_rpc).await?;
	Ok(())
}

#[cfg(feature = "test_validator")]
#[test_log::test(tokio::test)]
async fn config_update_authority_test_validator() -> anyhow::Result<()> {
	let compute_units = shared_config_update_authority_test(create_validator_rpc).await?;
	let rounded_compute_units = bitflip_program::round_compute_units_up(compute_units);

	check!(rounded_compute_units == 10_000);
	insta::assert_snapshot!(format!("{rounded_compute_units} CU"));
	shared::save_compute_units(
		"config_update_authority",
		compute_units,
		"Update the authority of the config",
	)?;

	Ok(())
}

#[test_log::test(tokio::test)]
async fn authority_must_change_test() -> anyhow::Result<()> {
	shared_authority_must_change_test(create_banks_client_rpc).await
}

#[cfg(feature = "test_validator")]
#[test_log::test(tokio::test)]
async fn authority_must_change_test_validator() -> anyhow::Result<()> {
	shared_authority_must_change_test(create_validator_rpc).await
}

async fn create_banks_client_rpc() -> anyhow::Result<impl ToRpcClient> {
	let provider = create_program_context_with_factory(|p| {
		let mut accounts = create_config_accounts();
		accounts.extend(create_token_accounts(false)?);

		for (key, account) in accounts {
			p.add_account(key, account.into());
		}

		Ok(())
	})
	.await?;

	Ok(provider)
}

#[cfg(feature = "test_validator")]
async fn create_validator_rpc() -> anyhow::Result<impl ToRpcClient> {
	let mut config_state_accounts = create_config_accounts();
	config_state_accounts.extend(create_token_accounts(false)?);

	let runner = shared::create_runner_with_accounts(config_state_accounts).await;

	Ok(runner)
}

async fn shared_config_update_authority_test<
	T: ToRpcClient,
	Fut: Future<Output = anyhow::Result<T>>,
	Create: FnOnce() -> Fut,
>(
	create_provider: Create,
) -> anyhow::Result<u64> {
	let provider = create_provider().await?;
	let rpc = provider.to_rpc();
	let authority_keypair = create_authority_keypair();
	let authority = authority_keypair.pubkey();
	let new_authority_keypair = Keypair::new();
	let new_authority = new_authority_keypair.pubkey();
	let config = get_pda_config().0;
	let recent_blockhash = rpc.get_latest_blockhash().await?;
	let ix = config_update_authority(&authority, &new_authority);
	let mut transaction =
		VersionedTransaction::new_unsigned_v0(&authority, &[ix], &[], recent_blockhash)?;

	let simulation = rpc.simulate_transaction(&transaction).await?;
	log::info!("simulation: {simulation:#?}");
	let compute_units = simulation.value.units_consumed.unwrap();

	transaction
		.try_sign(&[&new_authority_keypair], None)?
		.try_sign(&[&authority_keypair], None)?;

	let signature = rpc.send_and_confirm_transaction(&transaction).await?;
	rpc.confirm_transaction(&signature).await?;

	let config_state_data = rpc.get_account_data(&config).await?;
	let config_state_account = ConfigState::try_from_bytes(&config_state_data)?;
	let authority_redaction = create_insta_redaction(new_authority, "new_authority:pubkey");
	insta::assert_compact_json_snapshot!(config_state_account,{
		".authority" => insta::dynamic_redaction(authority_redaction),
	});

	Ok(compute_units)
}

async fn shared_authority_must_change_test<
	T: ToRpcClient,
	Fut: Future<Output = anyhow::Result<T>>,
	Create: FnOnce() -> Fut,
>(
	create_provider: Create,
) -> anyhow::Result<()> {
	let provider = create_provider().await?;
	let rpc = provider.to_rpc();
	let authority_keypair = create_authority_keypair();
	let authority = authority_keypair.pubkey();
	let recent_blockhash = rpc.get_latest_blockhash().await?;
	let ix = config_update_authority(&authority, &authority);
	let transaction =
		VersionedTransaction::new_unsigned_v0(&authority, &[ix], &[], recent_blockhash)?;
	let simulation = rpc.simulate_transaction(&transaction).await?;

	log::info!("simulation: {simulation:#?}");
	check!(
		simulation.value.err.unwrap()
			== TransactionError::InstructionError(
				0,
				InstructionError::Custom(BitflipError::DuplicateAuthority.into())
			)
	);

	Ok(())
}
