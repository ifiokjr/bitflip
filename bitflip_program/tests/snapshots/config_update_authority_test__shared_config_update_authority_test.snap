---
source: bitflip_program/tests/config_update_authority_test.rs
expression: config_state_account
snapshot_kind: text
---
{
  "version": 0,
  "authority": "[new_authority:pubkey]",
  "bump": 254,
  "treasuryBump": 255,
  "mintBitBump": 255,
  "mintKibibitBump": 255,
  "mintMebibitBump": 255,
  "mintGibibitBump": 255,
  "gameIndex": 0,
  "padding": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0
  ]
}
